File: fhir-search.py
# Author Joey Whelan
from argparse import ArgumentParser
from redis import Connection, from_url
from redis.commands.search.field import NumericField, TagField, TextField
from redis.commands.search.indexDefinition import IndexDefinition, IndexType
from redis.commands.search.query import Query
from redis.commands.search.aggregation import AggregateRequest, Desc, Asc
from redis.commands.search import reducers
import os
import json

FHIR_DIR = './synthea/output/fhir'
REDIS_URL = 'redis://localhost:12000'

class FHIR(object):
    def __init__(self, args):
        self.connection: Connection = from_url(args.url)
    
    def build_indices(self) -> None:
        print('*** Build indices ***')
        count = 0
        
        # location index
        idx_def = IndexDefinition(index_type=IndexType.JSON, prefix=['Location:'])
        schema = [  TagField('$.status', as_name='status'),
            TextField('$.name', as_name='name'),
            TagField('$.address.city', as_name='city'),
            TagField('$.address.state', as_name='state'),
            NumericField('$.position.longitude', as_name='longitude'),
            NumericField('$.position.latitude', as_name='latitude')
        ]
        self.connection.ft('location_idx').create_index(schema, definition=idx_def)
        count +=1

        # practionerRole index
        idx_def = IndexDefinition(index_type=IndexType.JSON, prefix=['PractitionerRole:'])
        schema = [  TextField('$.practitioner.display', as_name='physician'),
            TextField('$.specialty[*].text', as_name='specialty', sortable=True),
            TextField('$.location[*].display', as_name='location')
        ]
        self.connection.ft('practitionerRole_idx').create_index(schema, definition=idx_def)
        count += 1

        # medicationRequest index
        idx_def = IndexDefinition(index_type=IndexType.JSON, prefix=['MedicationRequest:'])
        schema = [  TagField('$.status', as_name='status'),
            TextField('$.medicationCodeableConcept.text', as_name='drug'),
            TextField('$.requester.display', as_name='prescriber', sortable=True),
            TextField('$.reasonReference[*].display', as_name='reason')
        ]
        self.connection.ft('medicationRequest_idx').create_index(schema, definition=idx_def)
        count += 1

        # immunization index
        idx_def = IndexDefinition(index_type=IndexType.JSON, prefix=['Immunization:'])
        schema = [  TextField('$.vaccineCode.text', as_name='vax', sortable=True),
            TextField('$.location.display', as_name='location'),
            TextField('$.occurrenceDateTime', as_name='date')
        ]
        self.connection.ft('immunization_idx').create_index(schema, definition=idx_def)
        count += 1

        # condition index
        idx_def = IndexDefinition(index_type=IndexType.JSON, prefix=['Condition:'])
        schema = [  TagField('$.clinicalStatus.coding[*].code', as_name='code'),
            TextField('$.code.text', as_name='problem'),
            TagField('$.recordedDate', as_name='date')
        ]
        self.connection.ft('condition_idx').create_index(schema, definition=idx_def)
        count += 1

        # claims index
        idx_def = IndexDefinition(index_type=IndexType.JSON, prefix=['Claim:'])
        schema = [  TagField('$.status', as_name='status'),
            TextField('$.insurance[*].coverage.display', as_name='insurer', sortable=True),
            NumericField('$.total.value', as_name='value')
        ]
        self.connection.ft('claims_idx').create_index(schema, definition=idx_def)
        count += 1

        print(f'{count} indices built.')

    def load_docs(self) -> None:
        print('\n*** Load documents ***')
        count = 0
        for file in os.listdir(FHIR_DIR):
            with open(os.path.join(FHIR_DIR, file), 'r') as infile:
                bundle = json.load(infile)
                for entry in bundle['entry']:
                    key = f'{entry["resource"]["resourceType"]}:{entry["resource"]["id"]}'
                    self.connection.json().set(key, '$', entry['resource'])
                    count += 1
        print(f'{count} documents loaded.')
    
    def run_scenarios(self) -> None:
        print('\n*** Search Scenarios ***')
        
        print('\n**** Location Scenario 1 ****')
        ''' Business Problem - Find 3 medical facilities in Alaska
        '''
        query = Query('(@status:{active} @state:{AK})').return_fields('$.name', '$.address.city').paging(0,3)
        result = self.connection.ft('location_idx').search(query)
        print(result.docs) 

        print('\n**** Location Scenario 2 ****')
        ''' Business Problem - Find the closest medical facility (in the database) to Woodland Park CO
        '''
        request = AggregateRequest('@status:{active}')\
        .load('@name', '@city', '@state', '@longitude', '@latitude')\
        .apply(meters='geodistance(@longitude, @latitude, -105.0569, 38.9939)')\
        .apply(miles='ceil(@meters*0.000621371)')\
        .sort_by(Asc('@miles'))\
        .limit(0,1)
        result = self.connection.ft('location_idx').aggregate(request)
        print(result.rows)

        print('\n**** PractionerRole Scenario 1 ****')
        ''' Business Problem - Find 3 General Practice physicians that work from a hospital
        '''
        query = Query('(@specialty:"General Practice" @location:hospital)').return_fields('$.practitioner.display').paging(0,3)
        result = self.connection.ft('practitionerRole_idx').search(query)
        print(result.docs) 

        print('\n**** PractitionerRole Scenario 2 ****')
        ''' Business Problem - Find the count of physicians per medical specialty
        '''
        request = AggregateRequest('*').group_by('@specialty', reducers.count().alias('count'))     
        result = self.connection.ft('practitionerRole_idx').aggregate(request)
        print(result.rows)

        print('\n**** MedicationRequest Scenario 1 ****')
        ''' Business Problem - Find the names of 3 medications that have been currently prescribed for bronchitis
        '''
        query = Query('(@status:{active} @reason:%bronchitis%)').return_fields('$.medicationCodeableConcept.text').paging(0,3)
        result = self.connection.ft('medicationRequest_idx').search(query)
        print(result.docs)

        print('\n**** MedicationRequest Scenario 2 ****')
        ''' Business Problem - Find the top 3 physicians by prescription count who are prescribing opioids
        '''
        opioids = 'Hydrocodone|Oxycodone|Oxymorphone|Morphine|Codeine|Fentanyl|Hydromorphone|Tapentadol|Methadone'
        request = AggregateRequest(f'@drug:{opioids}')\
        .group_by('@prescriber', reducers.count().alias('opioids_prescribed'))\
        .sort_by(Desc('@opioids_prescribed'))\
        .limit(0,3)
        result = self.connection.ft('medicationRequest_idx').aggregate(request)
        print(result.rows)

        print('\n**** Immunization Scenario 1 ****')
        '''  Business Problem - Find 5 patients that received an immunizations at an urgent care clinic in 2015
        '''
        query = Query('@location:urgent @date:2015*').return_fields('$.patient.reference').paging(0,5)
        result = self.connection.ft('immunization_idx').search(query)
        print(result.docs)

        print('\n**** Immunization Scenario 2 ****')
        '''  Business Problem - Find the top 5 vaccines administered in 2020.
        '''
        request = AggregateRequest('@date:2020*')\
        .group_by('@vax', reducers.count().alias('num_vax'))\
        .sort_by(Desc('@num_vax'))\
        .limit(0,5)
        result = self.connection.ft('immunization_idx').aggregate(request)
        print(result.rows)

        print('\n**** Condition Scenario 1 ****')
        '''  Business Problem - Find 3 patients with active cases of rhinitis or asthma
        '''
        query = Query('@code:{active} @problem:(rhinitis|asthma)').return_fields('$.subject.reference').paging(0, 3)
        result = self.connection.ft('condition_idx').search(query)
        print(result.docs)

        print('\n**** Condition Scenario 2 ****')
        ''' Business Problem - Find the count of reported medical conditions categorized by year
        '''
        request = AggregateRequest('*')\
        .load('@date')\
        .apply(year='substr(@date,0,4)')\
        .group_by('@year', reducers.count().alias('num_conditions'))\
        .sort_by(Desc('@year'))\
        .limit(0,5)
        result = self.connection.ft('condition_idx').aggregate(request)
        print(result.rows)

        print('\n**** Claims Scenario 1 ****')
        ''' Business Problem - Find the service description of 3 active claims where Aetna is the insurer and the claim value is greater than $1000
        '''
        query = Query('@insurer:Aetna @value:[1000,+inf] @status:{active}').return_fields('$.item[0].productOrService.text').paging(0, 3)
        result = self.connection.ft('claims_idx').search(query)
        print(result.docs)

        print('\n**** Claims Scenario 2 ****')
        ''' Business Problem - Find the Top 3 insurers by claim value
        '''
        request = AggregateRequest('@status:{active}')\
        .group_by('@insurer', reducers.sum('@value').alias('total_value'))\
        .filter('@total_value > 0')\
        .sort_by(Desc('@total_value'))\
        .limit(0,3)
        result = self.connection.ft('claims_idx').aggregate(request)
        print(result.rows)



if __name__ == '__main__':
    parser = ArgumentParser(description='FHIR Search Examples')
    parser.add_argument('--url', required=False, type=str, default=REDIS_URL,
        help='Redis URL connect string')
    args = parser.parse_args()
    fhir = FHIR(args)
    fhir.build_indices()
    fhir.load_docs()
    fhir.run_scenarios()
    

File: fhir-search.js
/**
 * @maker Joey Whelan
 * @fileoverview Series of Redis Search and Aggregate scenarios
 */

import { AggregateGroupByReducers, AggregateSteps, createClient, SchemaFieldTypes } from 'redis';
import { ArgumentParser } from 'argparse';
import fsPromises from 'node:fs/promises';
const FHIR_DIR = './synthea/output/fhir'
const REDIS_URL = 'redis://localhost:12000'

class FHIR {
    constructor(args) {
        this.url = args.url;
    }

    async connect() {
        this.client = createClient({url: this.url});
        this.client.on('error', (err) => {
            console.error(err.message);
        });  
        await this.client.connect();
    }

    async disconnect() {
        await this.client.disconnect();
    }

    async buildIndices() {
        console.log('*** Build indices ***');
        let count = 0;

        // location index
        await this.client.ft.create('location_idx', {
            '$.status': {
                type: SchemaFieldTypes.TAG,
                AS: 'status'  
            },
            '$.name': {
                type: SchemaFieldTypes.TEXT,
                AS: 'name'
            },
            '$.address.city': {
                type: SchemaFieldTypes.TAG,
                AS: 'city'
            },
            '$.address.state': {
                type: SchemaFieldTypes.TAG,
                AS: 'state'
            },
            '$.position.longitude': {
                type: SchemaFieldTypes.NUMERIC,
                AS: 'longitude'
            },
            '$.position.latitude': {
                type: SchemaFieldTypes.NUMERIC,
                AS: 'latitude'
            }
        }, { ON: 'JSON', PREFIX: 'Location:'});
        count++;

        // practitionerRole index
        await this.client.ft.create('practitionerRole_idx', {
            '$.practitioner.display': {
                type:  SchemaFieldTypes.TEXT,
                AS: 'physician'
            },
            '$.specialty[*].text': {
                type: SchemaFieldTypes.TEXT,
                AS: 'specialty',
                SORTABLE: true
            },
            '$.location[*].display': {
                type: SchemaFieldTypes.TEXT,
                AS: 'location'
            }
        }, {ON: 'JSON', PREFIX: 'PractitionerRole:'});
        count++;

        // medicationRequest index
        await this.client.ft.create('medicationRequest_idx', {
            '$.status': {
                type: SchemaFieldTypes.TAG,
                AS: 'status'
            },
            '$.medicationCodeableConcept.text': {
                type: SchemaFieldTypes.TEXT,
                AS: 'drug'
            },
            '$.requester.display': {
                type: SchemaFieldTypes.TEXT,
                AS: 'prescriber',
                SORTABLE: true
            },
            '$.reasonReference[*].display': {
                type: SchemaFieldTypes.TEXT,
                AS: 'reason'
            }
        }, {ON: 'JSON', PREFIX: 'MedicationRequest:'});
        count++;

        // immunization index
        await this.client.ft.create('immunization_idx', {
            '$.vaccineCode.text': {
                type: SchemaFieldTypes.TEXT,
                AS: 'vax',
                SORTABLE: true
            },
            '$.location.display': {
                type: SchemaFieldTypes.TEXT,
                AS: 'location'
            },
            '$.occurrenceDateTime': {
                type: SchemaFieldTypes.TEXT,
                AS: 'date'
            }
        }, {ON: 'JSON', PREFIX: 'Immunization:'});
        count++;

        // condition index
        await this.client.ft.create('condition_idx', {
            '$.clinicalStatus.coding[*].code': {
                type: SchemaFieldTypes.TAG,
                AS: 'code'
            },
            '$.code.text': {
                type: SchemaFieldTypes.TEXT,
                AS: 'problem'
            },
            '$.recordedDate': {
                type: SchemaFieldTypes.TAG,
                AS: 'date'
            }
        }, {ON: 'JSON', PREFIX: 'Condition:'});
        count++;

        // claims index
        await this.client.ft.create('claims_idx', {
            '$.status': {
                type: SchemaFieldTypes.TAG,
                AS: 'status'
            },
            '$.insurance[*].coverage.display': {
                type: SchemaFieldTypes.TEXT,
                AS: 'insurer',
                SORTABLE: true    
            },
            '$.total.value': {
                type: SchemaFieldTypes.NUMERIC,
                AS: 'value'
            }
        }, {ON: 'JSON', PREFIX: 'Claim:'});
        count++;
        console.log(`${count} indices built.`);
    }

    async loadDocs() {
        console.log('\n*** Load documents ***')
        let count = 0;
        let files = await fsPromises.readdir(FHIR_DIR);
        files = files.filter(file => file.endsWith(('.json')));
        for (const file of files) {
            const data = await fsPromises.readFile(`${FHIR_DIR}/${file}`);
            const bundle = JSON.parse(data);
            for (const entry of bundle.entry) {
                const key = `${entry.resource.resourceType}:${entry.resource.id}`  
                await this.client.json.set(key, '.', entry.resource);  
                count++;  
            }
        }   
        console.log(`${count} documents loaded.`);
    }

    async runScenarios() {
        console.log('\n*** Search Scenarios ***');
        
        console.log('\n**** Location Scenario 1 ****');
        // Business Problem - Find 3 medical facilities in Alaska
        let result = await this.client.ft.search('location_idx', '(@status:{active} @state:{AK})', {
            RETURN: ['$.name', '$.address.city'],
            LIMIT: {from: 0, size: 3}
        });
        console.log(JSON.stringify(result.documents,null,4));

        console.log('\n**** Location Scenario 2 ****');
        // Business Problem - Find the closest medical facility (in the database) to Woodland Park CO
        result = await this.client.ft.aggregate('location_idx','@status:{active}', {
            LOAD: ['@name', '@city', '@state', '@longitude', '@latitude'],
            STEPS: [
                    {   type: AggregateSteps.APPLY,
                        expression: 'geodistance(@longitude, @latitude, -105.0569, 38.9939)', 
                        AS: 'meters' 
                    },
                    {   type: AggregateSteps.APPLY ,
                        expression: 'ceil(@meters*0.000621371)', 
                        AS: 'miles' 
                    },
                    {
                        type: AggregateSteps.SORTBY,
                        BY: {
                            BY: '@miles', 
                            DIRECTION: 'ASC' 
                        }
                    },
                    {
                        type: AggregateSteps.LIMIT,
                        from: 0, 
                        size: 1
                    }
            ]
        });
        console.log(JSON.stringify(result.results,null,4));

        console.log('\n**** PractionerRole Scenario 1 ****')
        // Business Problem - Find 3 General Practice physicians that work from a hospital
        result = await this.client.ft.search('practitionerRole_idx', '(@specialty:"General Practice" @location:hospital)', {
            RETURN: ['$.practitioner.display'],
            LIMIT: {from: 0, size: 3}
        });
        console.log(JSON.stringify(result.documents,null,4));

        console.log('\n**** PractitionerRole Scenario 2 ****');
        // Business Problem - Find the count of physicians per medical specialty
        result = await this.client.ft.aggregate('practitionerRole_idx', '*', {
            STEPS: [
                {   type: AggregateSteps.GROUPBY,
                    properties: ['@specialty'],
                    REDUCE: [
                        {   type: AggregateGroupByReducers.COUNT,
                            property: '@speciality',
                            AS: 'count'
                        }
                    ]   
                }
            ]
        })
        console.log(JSON.stringify(result.results,null,4));

        console.log('\n**** MedicationRequest Scenario 1 ****')
        // Business Problem - Find the names of 3 medications that have been currently prescribed for bronchitis
        result = await this.client.ft.search('medicationRequest_idx', '(@status:{active} @reason:%bronchitis%)', {
            RETURN: ['$.medicationCodeableConcept.text'],
            LIMIT: {from: 0, size: 3}
        });
        console.log(JSON.stringify(result.documents,null,4));

        console.log('\n**** MedicationRequest Scenario 2 ****')
        // Business Problem - Find the top 3 physicians by prescription count who are prescribing opioids
        const opioids = 'Hydrocodone|Oxycodone|Oxymorphone|Morphine|Codeine|Fentanyl|Hydromorphone|Tapentadol|Methadone';
        result = await this.client.ft.aggregate('medicationRequest_idx', `@drug:${opioids}`, {
            STEPS: [
                {   type: AggregateSteps.GROUPBY,
                    properties: ['@prescriber'],
                    REDUCE: [
                        {   type: AggregateGroupByReducers.COUNT,
                            property: '@prescriber',
                            AS: 'opioids_prescribed'
                        }
                    ]   
                },
                {
                    type: AggregateSteps.SORTBY,
                    BY: { 
                        BY: '@opioids_prescribed', 
                        DIRECTION: 'DESC' 
                    }
                },
                {
                    type: AggregateSteps.LIMIT,
                    from: 0, 
                    size: 3
                }
            ]
        });
        console.log(JSON.stringify(result.results,null,4));

        console.log('\n**** Immunization Scenario 1 ****')
        // Business Problem - Find 5 patients that received an immunizations at an urgent care clinic in 2015
        result = await this.client.ft.search('immunization_idx', '@location:urgent @date:2015*', {
            RETURN: ['$.patient.reference'],
            LIMIT: {from: 0, size: 5}
        });
        console.log(JSON.stringify(result.documents,null,4));


        console.log('\n**** Immunization Scenario 2 ****')
        // Business Problem - Find the top 5 vaccines administered in 2020.
        result = await this.client.ft.aggregate('immunization_idx', '@date:2020*', {
            STEPS: [{   
                type: AggregateSteps.GROUPBY,
                properties: ['@vax'],
                REDUCE: [{   
                    type: AggregateGroupByReducers.COUNT,
                    property: '@vax',
                    AS: 'num_vax'
                }]},
                {   type: AggregateSteps.SORTBY,
                    BY: { 
                    BY: '@num_vax', 
                    DIRECTION: 'DESC' 
                }},
                {   type: AggregateSteps.LIMIT,
                    from: 0, 
                    size: 5
                }
            ]
        });
        console.log(JSON.stringify(result.results,null,4));

        console.log('\n**** Condition Scenario 1 ****')
        // Business Problem - Find 3 patients with active cases of rhinitis or asthma
        result = await this.client.ft.search('condition_idx', '@code:{active} @problem:(rhinitis|asthma)', {
            RETURN: ['$.subject.reference'],
            LIMIT: {from: 0, size: 3}
        });
        console.log(JSON.stringify(result.documents,null,4));

        console.log('\n**** Condition Scenario 2 ****')
        // Business Problem - Find the count of reported medical conditions categorized by year
        result = await this.client.ft.aggregate('condition_idx', '*', {
            LOAD: ['@date'],
            STEPS: [
                {   type: AggregateSteps.APPLY,
                    expression: 'substr(@date,0,4)', 
                    AS: 'year' 
                },   
                {   type: AggregateSteps.GROUPBY,
                    properties: ['@year'],
                    REDUCE: [{   
                        type: AggregateGroupByReducers.COUNT,
                        property: '@year',
                        AS: 'num_conditions'
                }]},
                {   type: AggregateSteps.SORTBY,
                    BY: { 
                    BY: '@year', 
                    DIRECTION: 'DESC' 
                }},
                {   type: AggregateSteps.LIMIT,
                    from: 0, 
                    size: 5
                }
            ]
        });
        console.log(JSON.stringify(result.results,null,4));

        console.log('\n**** Claims Scenario 1 ****')
        // Business Problem - Find the service description of 3 active claims where Aetna is the insurer and the claim value is greater than $1000
        result = await this.client.ft.search('claims_idx', '@insurer:Aetna @value:[1000,+inf] @status:{active}', {
            RETURN: ['$.item[0].productOrService.text'],
            LIMIT: {from: 0, size: 3}
        });
        console.log(JSON.stringify(result.documents,null,4));

        console.log('\n**** Claims Scenario 2 ****')
        // Business Problem - Find the Top 3 insurers by claim value
        result = await this.client.ft.aggregate('claims_idx', '@status:{active}', {
            STEPS: [
                {   type: AggregateSteps.GROUPBY,
                    properties: ['@insurer'],
                    REDUCE: [{   
                        type: AggregateGroupByReducers.SUM,
                        property: '@value',
                        AS: 'total_value'
                }]},
                {
                    type: AggregateSteps.FILTER,
                    expression: '@total_value > 0'
                },
                {   type: AggregateSteps.SORTBY,
                    BY: { 
                    BY: '@total_value', 
                    DIRECTION: 'DESC' 
                }},
                {   type: AggregateSteps.LIMIT,
                    from: 0, 
                    size: 5
                }
            ]
        });
        console.log(JSON.stringify(result.results,null,4));
    }
}

(async () => {
    
    const parser = ArgumentParser({ description: 'FHIR Search Examples' });
    parser.add_argument('--url', {
            required: false,
            type: 'str',
            help: 'Redis URL connect string',
            default: REDIS_URL
    });
    const args = parser.parse_args();
    const fhir = new FHIR(args);
    await fhir.connect();
    await fhir.buildIndices();
    await fhir.loadDocs();
    await fhir.runScenarios();
    await fhir.disconnect();
})();